### PR TITLE
DataFlow: Add language-specific predicate for ignoring steps in flow-through calculation

### PR DESCRIPTION
**Pull request by MathiasVP**
### TLDR

This PR adds a language-level hook for excluding steps from the flow-through calculation. It's a no-op PR for all other languages than C/C++.

### Slightly longer explanation

We have a well-known source of FPs in C/C++ caused by the following pattern:
```cpp
void modify(int* q) { *q = source(); }

void modify_copy(int* p) {
  int x = *p;
  modify(&x);
}

void test() {
  int x;
  modify_copy(&x);
  sink(x); // we report flow from source to sink here.
}
```

This happens because we model each indirection of a parameter as a separate SSA variable, so dataflow really sees the above program as
```cpp
void modify(int* q, int deref_q) { deref_q = source(); }

void modify_copy(int* p, int deref_p) {
  int x = deref_p;
  modify(&x, x);
}

void test() {
  int x;
  modify_copy(&x, x);
  sink(x); // we report flow from source to sink here.
}
```
and in the above program it's easy to see that there's a `ReturnNodeExt` generated that updates the parameter `deref_p` in `modify_copy`.

The problem is the SSA step generated by `int x = *p;` which should be treated special when generating flow-throughs since a subsequent update of `x` should not affect `*p` (which is currently what the flow-through summaries give us).

@aschackmull I'd like to hear if you have any problems with 748625c7e5e5253da49a68524df53a5af171168e? I hope it's not controversial since:
- it's a no-op for all the other languages
- it's a very simple fix for a problem that's been observed quite often in C/C++

DCA looks good. I've verified that all the removed results are instances of the `test_modify_copy_of_pointer` testcase.

### CODE REVIEW

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/flowOut.cpp [hunk 1]
```diff
@@ -1,20 +1,74 @@
...
+void modify(int* ptr) { // $ ast-def=ptr
+  int* deref = ptr;
+  int* other = &*deref;
+  source_ref(other);
+}
+
+void test_output() {
+   int x = 0;
+   modify(&x);
+   sink(x); // $ SPURIOUS: ir MISSING: ast
```
Comment (aschackmull) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/flowOut.cpp:31]: There's supposed to be flow here, right? I.e. it shouldn't have a `SPURIOUS` label.

Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/flowOut.cpp:31]: Oops, you're right. That was a copy/paste fail 😅

File: shared/dataflow/codeql/dataflow/DataFlow.qll [hunk 2]
```diff
@@ -170,6 +170,13 @@ signature module InputSig {
 
   predicate simpleLocalFlowStep(Node node1, Node node2);
 
+  /**
+   * Holds if the data flow step from `node1` to `node2` can be used when
+   * computing flow-through summaries.
+   */
+  bindingset[node1, node2]
+  default predicate flowThroughStepAllowed(Node node1, Node node2) { any() }
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply shared/dataflow/codeql/dataflow/DataFlow.qll]: Good idea. Renamed in 816adfabdedca8b0e8c92fdffe32c7b47d006868.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/flowOut.cpp (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/flowOut.cpp]: Fixed in 5d9973bb57d99ac8fd80a64d11260a7219d26942.

File: shared/dataflow/codeql/dataflow/DataFlow.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply shared/dataflow/codeql/dataflow/DataFlow.qll]: Oh, yes. Would you prefer something along the lines of:
```ql
/**
 * Holds if the data-flow step from `node1` to `node2` can be used to
 * determine where side-effects may return from a callable.
 */
```
?

(I figured I might as well steal your phrasing 😂)

[the remark this answers is not in the document]
Comment (MathiasVP) [reply shared/dataflow/codeql/dataflow/DataFlow.qll]: I went with the above in 911f1543e02cee987c451eb3063ef6da9dfac77a. Let me know if you prefer something else.

**Review by aschackmull — APPROVED** [review]
